timemachine/fe/free_energy.py

```diff
@@ -1120,6 +1136,91 @@ def compute_dense():
     return U_kl
 
 
+def make_u_kl_fxn(trajs, initial_states):
+    """fxn(k, l) = "trajs[k] evaluated in ensembles[l]"
+
+    usage note: be careful of axis-ordering convention, see: https://github.com/proteneer/timemachine/issues/1100
+    """
+
+    summed_potentials = [make_summed_potential(s.potentials) for s in initial_states]
+
+    # validate assumption that initial states all have compatible potentials / ensembles
+    kBTs = [BOLTZ * state.integrator.temperature for state in initial_states]
+    s_0 = initial_states[0]
+    for s in initial_states[1:]:
+        assert_ensembles_compatible(s_0, s)
+        assert_potentials_compatible(s_0.potentials, s.potentials)
+
+    sp = summed_potentials[0]
+    sp_gpu = sp.potential.to_gpu(np.float32)
+
+    def batch_U_fxn(xs, ps, bs, x_idxs, p_idxs):
+        Us = sp_gpu.unbound_impl.execute_batch_sparse(xs, ps, bs, x_idxs, p_idxs, False, False, True)[2]
+        return np.nan_to_num(Us, nan=+np.inf)
+
+    def u_kl(k, l):
+        coords = trajs[k].frames
+        boxes = trajs[k].boxes
+
+        params = np.array([summed_potentials[l].params])
```

Comment:
nit: `np.asarray` to avoid copy

Reply:
Will leave this as-is for simplicity (don't have to think about possibility of accidental in-place mutation of any `summed_potential`), unless the copy is costly

Reply:
Makes sense; I'd misunderstood the intent as ensuring that `params` is a numpy array

Reply:
That is the main intent, and another intent is to make it have shape `(1, len(summed_potentials[l].params))`